test/functional/sessions.test.js

```diff
@@ -123,7 +123,7 @@ describe('Sessions', function () {
               // verify that the `endSessions` command was sent
               const lastCommand = test.commands.started[test.commands.started.length - 1];
               expect(lastCommand.commandName).to.equal('endSessions');
-              expect(client.topology.s.sessionPool.sessions).to.have.length(0);
+              expect(client.topology).to.not.exist;
```

Comment:
As part of closing the `MongoClient`, we set `client.topology` to `undefined`, so we cannot check the size of its session pool, we can just check that the topology no longer exists. 

I agree it's not the best but it seems like we _have_ to set the topology to `undefined` (instead of just setting the state of the topology to closed) because otherwise it's ambiguous whether the topology 1) just hasn't been opened yet or 2) was closed/all cleaned up.

Reply:
Hmm. I think an implicit part of this work is that `Topology` can be connected, closed and reused as well. A `Topology` basically _is_ a client, the `MongoClient` being a wrapper around it to provide access to higher level API. Ideally we would want to assign a `Topology` in the constructor of `MongoClient`, rather than during `connect` (but that's not quite possible today).

Have you tried not resetting `topology = undefined` yet? This might require changes to the `connect` operation. Let's not go down a huge rabbit hole here, but I'd like to better understand the limitations to that approach before opting for this alternative.

Reply:
Yep, you're right, it is possible to _not_ set `topology = undefined` after a small change to the `connect` logic. This means we don't need to do any additional clean-up on client close, since any references to a `Db`/`Client` can use the same topology after re-connect. I've made this change, and will write a few tests to verify that the topology still receives the correct events after client re-connect.